logstash-core/src/main/java/org/logstash/execution/inputs/HttpPoller.java

```diff
@@ -0,0 +1,63 @@
+package org.logstash.execution.inputs;
+
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Map;
+import org.logstash.execution.Input;
+import org.logstash.execution.LsConfiguration;
+import org.logstash.execution.LsContext;
+import org.logstash.execution.PluginConfigSpec;
+import org.logstash.execution.QueueWriter;
+
+public final class HttpPoller implements Input {
+
+    private static final PluginConfigSpec<String> USER_CONFIG =
+        LsConfiguration.stringSetting("user");
+
+    private static final PluginConfigSpec<String> PASSWORD_CONFIG =
+        LsConfiguration.stringSetting("password");
+
+    private static final PluginConfigSpec<Long> AUTOMATIC_RETRIES_CONFIG =
+        LsConfiguration.numSetting("automatic_retries", 1L);
+
+    private static final PluginConfigSpec<Path> CA_CERT_CONFIG =
+        LsConfiguration.pathSetting("cacert");
+
+    private static final PluginConfigSpec<Map<String, String>> URLS_CONFIG =
```

Comment:
This doesn't really escapsulate the urls setting in the poller. The value can be either a `String`, as it is here, or a complex hash: https://www.elastic.co/guide/en/logstash/current/plugins-inputs-http_poller.html . That was my reason for us trying to figure out that case here.

My suspicion is that we'll wind up with `Map<String, Object>`, and use code to sort out the mess.

I have no idea how to make this programmatic for a UI.

@ycombinator , your thoughts?

Reply:
Reading https://www.elastic.co/guide/en/logstash/current/plugins-inputs-http_poller.html#plugins-inputs-http_poller-urls, it looks like this is always expected to be a hash? The `Map<String, String>` type in the code on this line seems to mirror that expectation too. I guess I'm not seeing how this could be set as "just" a single `String`?

> I have no idea how to make this programmatic for a UI.

One way to encode this in a spec for the UI would be to use some sort of a special designation for user-defined keys in a hash/object data type. Let me try to elaborate below.

First, consider a setting whose datatype is a hash, but with keys provided by the plugin itself. For example, the [`schedule` setting](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-http_poller.html#plugins-inputs-http_poller-schedule). In this case, the spec for this setting could be something like the following (stealing from ES mapping syntax for illustration purposes only):

```
{
   "schedule": {
     "type": "object" // or "hash" if you prefer that
     "properties": {
       "cron": { "type": "string" },
       "every": { "type": "string" },
       "in": { "type": "string" },
       "at": { "type": "string" }
     }
  }
}
```

Now, consider the `urls` setting. The spec for that could be:

```
{
   "schedule": {
     "type": "object" // or "hash" if you prefer that
     "properties": {
       "__user_defined__": { "type": "string" 
     }
  }
}
```

Thoughts?

Reply:
@ycombinator @andrewvc my thought was that we could programmatically encapsulate what ES does as `PluginConfigSpec<Map<String, PluginConfigSpec>>` and build up a nested mapping that way.
For some reason, I forgot adding that part ... will be up soon.

Reply:
@ycombinator it can be either a string or a hash. It's weird that way.

@original-brownbear ah, yes, that makes more sense.

Still, encoding this sort of thing into JSON is inherently strange. It's pretty complex to encode can be a string or a map, and even after you do encode that, how do you render the UI?

So, here's a thought, maybe there's some sort of 'override' UI spec the plugin can bundle. In this case, we don't need the UI to support the simple syntax, just the full one. We could let the plugin author manually specify what the UI controls should be instead of doing that programatically.

Reply:
@andrewvc I'm good with that proposal — that the UI gets the "highest fidelity" syntax. We can always figure out optimizations/shortcuts later.

WDYT about my proposal for encoding plugin-specified keys vs. user-defined keys above?

Reply:
@andrewvc here you go, quick sketch but the idea should be clear:

https://github.com/elastic/logstash/pull/9137/commits/11f27182bd11a8776302b2b8063bfd23e522ad87

I think you can simply encode this as a nesting of `LsConfiguration`. This way the UI can also render it as such with the ability to add and remove elements (that are essentially the same as the top-level element).

In this case, you could simply add a UI element that allows adding nested "rows" for each `Map<String, LsConfiguration>` typed setting.
Hopefully, this makes sense, happy to go into more detail, but I'm optimistic we can properly encode this for the UI :)

Reply:
@ycombinator I tried to capture your proposal in some Java code.

I actually left out the spot with the flat hash I just noticed, my bad:

```
    @SuppressWarnings("unchecked")
    public static <T> PluginConfigSpec<Map<String, T>> requiredFlatHashSetting(
        final String name, Class<T> type) {
        //TODO: enforce subtype
        return new PluginConfigSpec(
            name, Map.class, null, false, true
        );
    }

    @SuppressWarnings("unchecked")
    public static PluginConfigSpec<Map<String, LsConfiguration>> requiredNestedHashSetting(
        final String name, final Collection<PluginConfigSpec<?>> spec) {
        return new PluginConfigSpec(
            name, Map.class, null, false, true, spec
        );
    }
```

Something like this would be doable though that captures flat sub-hashes :)
I think we can easily render that in that or a similar ES style JSON for the UI.

Reply:
@original-brownbear I have to admit, I'm a bit lost :) But that's okay! I mostly just want to know, from a UI POV, if we'll be able to encode the following:

1. A setting whose data type is a hash. Almost 100% certain this is possible :)

2. For such settings, there are two cases for the hash's keys:
    a. Plugin-defined, like in the case of the [`schedule` setting](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-http_poller.html#plugins-inputs-http_poller-schedule) setting, and
    b. User-defined, like in the case of the [`urls` setting](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-http_poller.html#plugins-inputs-http_poller-urls)

    The UI needs to be able to distinguish between these two cases by looking at the spec so it can decide whether to a) show "static" keys or b) make the key a text box field for the user to type something in, and also allow them to add N number of such rows.

    So my question is, can this sort of distinction be generated from the Java API you are proposing?

Reply:
@ycombinator yes, that's doable and was exactly the plan in the Java snippet above :)